prow/gcs/reporter/reporter.go

```diff
@@ -0,0 +1,253 @@
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/testgrid/metadata"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/googleapi"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	prowv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/gcsupload"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+)
+
+const reporterName = "gcsreporter"
+
+type gcsReporter struct {
+	cfg           config.Getter
+	dryRun        bool
+	logger        *logrus.Entry
+	podClientSets map[string]corev1.CoreV1Interface
+	storage       *storage.Client
+}
+
+func (gr *gcsReporter) Report(pj *prowv1.ProwJob) ([]*prowv1.ProwJob, error) {
+	ctx := context.Background() // TODO: get this from somewhere better?
+
+	_, _, err := gr.getJobDestination(pj)
+	if err != nil {
+		return []*prowv1.ProwJob{pj}, nil
+	}
+	// TODO: do something with these errors?
+	// TODO: parallelise?
+	_ = gr.reportPodInfo(pj, ctx)
+	_ = gr.reportFinishedJob(pj, ctx)
+	_ = gr.reportProwjob(pj, ctx)
+	return []*prowv1.ProwJob{pj}, nil
+}
+
+// reportPodInfo reports information about the pod that the job ran in. In particular,
+// it reports the pod status, as well as any events concerning the pod.
+func (gr *gcsReporter) reportPodInfo(pj *prowv1.ProwJob, ctx context.Context) error {
+	// We only report this after a prowjob is complete (and, therefore, pod state is immutable)
+	if !pj.Complete() {
+		return nil
+	}
+
+	pod, err := gr.podClientSets[pj.ClusterName].Pods(gr.cfg().PodNamespace).Get(pj.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	var events *v1.EventList
+	if pod != nil {
+		// we don't particularly care if this fails.
+		events, _ = gr.podClientSets[pj.ClusterName].Events(gr.cfg().PodNamespace).Search(scheme.Scheme, pod)
+	}
+
+	outputMap := map[string]interface{}{}
+	if pod != nil {
+		outputMap["pod"] = pod
+	}
+	if events != nil && len(events.Items) > 0 {
+		outputMap["events"] = events.Items
+	}
+
+	output, err := json.Marshal(outputMap)
+	if err != nil {
+		// This should never happen.
+		gr.logger.WithError(err).Warn("Couldn't marshal pod info")
+	}
+
+	bucketName, dir, err := gr.getJobDestination(pj)
+	if err != nil {
+		return err
+	}
+
+	if gr.dryRun {
+		gr.logger.Infof("Would upload pod info to %q/%q", bucketName, dir)
+		return nil
+	}
+
+	b := gr.storage.Bucket(bucketName)
+	w := b.Object(path.Join(dir, "pod_info.json")).NewWriter(ctx)
+	_, err = w.Write(output)
+	if err != nil {
+		gr.logger.WithError(err).Warn("Uploading pod info to GCS failed (write)")
+	}
+	err = w.Close()
+	if err != nil {
+		gr.logger.WithError(err).Warn("Uploading pod info to GCS failed (close)")
+	}
+	return err
+}
+
+// reportFinishedJob uploads a finished.json for the job, iff one did not already exist.
+func (gr *gcsReporter) reportFinishedJob(pj *prowv1.ProwJob, ctx context.Context) error {
```

Comment:
Nit: Per convention, `context.Context` is always the first arg to a function if present